File: Java-CodeServer/src/main/resources/js/prompto/literal/TextLiteral.js
var Literal = require("./Literal").Literal;
var Text = require("../value/Text").Text;
var TextType = require("../type/TextType").TextType;

/*jshint evil:true*/
function unescape(text) {
	return eval(text);
}

function TextLiteral(text) {
	Literal.call(this, text, new Text(unescape(text)));
	return this;
}

TextLiteral.prototype = Object.create(Literal.prototype);
TextLiteral.prototype.constructor = TextLiteral;


TextLiteral.prototype.check = function(context) {
	return TextType.instance;
};

exports.TextLiteral = TextLiteral;

File: Core/src/main/java/prompto/literal/TextLiteral.java
package prompto.literal;

import java.io.IOException;
import java.io.StreamTokenizer;
import java.io.StringReader;

import prompto.compiler.Flags;
import prompto.compiler.IConstantOperand;
import prompto.compiler.MethodInfo;
import prompto.compiler.Opcode;
import prompto.compiler.ResultInfo;
import prompto.compiler.StringConstant;
import prompto.runtime.Context;
import prompto.transpiler.Transpiler;
import prompto.type.IType;
import prompto.type.TextType;
import prompto.value.Text;


public class TextLiteral extends Literal<Text> {

	public TextLiteral(String text) {
		super(text, unescape(text));
	}

	private static Text unescape(String text) {
		StreamTokenizer parser = new StreamTokenizer(new StringReader(text));
		try {
		  parser.nextToken();
		  return new Text(parser.sval);
		}
		catch (IOException e) {
		  throw new RuntimeException(e);
		}
	}

	@Override
	public IType check(Context context) {
		return TextType.instance();
	}
	

	@Override
	public ResultInfo compile(Context context, MethodInfo method, Flags flags) {
		IConstantOperand operand = new StringConstant(value.getStorableData());
		method.addInstruction(Opcode.LDC_W, operand);
		return new ResultInfo(String.class);
	}
	
	@Override
	public void declare(Transpiler transpiler) {
		// nothing to do
	}
	
	@Override
	public boolean transpile(Transpiler transpiler) {
		transpiler.append(this.text.get());
		return false;
	}
}
